torch/optim/sgd.py

```diff
@@ -349,8 +355,10 @@ def _single_tensor_sgd(
                 grad = grad.add(buf, alpha=momentum)
             else:
                 grad = buf
-
-        param.add_(grad, alpha=-lr)
+        if differentiable and isinstance(lr, Tensor) and lr.requires_grad:
+            param.addcmul_(grad, -lr.to(param.device))
+        else:
+            param.add_(grad, alpha=-lr)
```

Comment:
Does lr as a tensor when SGD is not differentiable work here? If not, then we can relax the above conditional to just be if lr is a Tensor to increase support.

Later, indeed, this should lead to widening the test cases in common_optimizers.py for SGD, and can be done in another PR if python test/test_optim.py -k SGD doesn't automatically pass (it might not pass for foreach/fused and noncapturable).